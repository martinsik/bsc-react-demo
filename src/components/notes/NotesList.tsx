import React, { useState } from 'react';
import { Observable } from 'rxjs';
import { Table, OverlayTrigger, Tooltip } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
// @ts-ignore
import FaTrash from 'react-icons/lib/fa/trash';
// @ts-ignore
import FaEdit from 'react-icons/lib/fa/edit';

import { Note } from '../../model/Note';
import { RemoveNoteModal } from './RemoveNoteModal';
import { createNoteDetailRoute, createNoteEditRoute } from '../../routes';

import styles from './NotesList.module.scss'

interface NotesListProps {
  notes: Note[];
  onRemoveNote: (note: Note) => Observable<any>;
}

interface NoteRowProps {
  note: Note;
  onRemove: () => void;
}

const NoteRow = (props: NoteRowProps) => {
  const { note, onRemove } = props;

  return (
    <tr>
      <td>{note.id}</td>
      <td>
        <Link to={createNoteDetailRoute(note.id)}>{note.title}</Link>
      </td>
      <td>
        <OverlayTrigger
          placement="top"
          overlay={<Tooltip id={`note-tooltip-edit-${note.id}`}><FormattedMessage id="notes.list.edit" /></Tooltip>}
        >
          <Link to={createNoteEditRoute(note.id)}>
            <FaEdit />
          </Link>
        </OverlayTrigger>
        <OverlayTrigger
          placement="top"
          overlay={<Tooltip id={`note-tooltip-remove-${note.id}`}><FormattedMessage id="notes.list.remove" /></Tooltip>}
        >
          <a href="/" onClick={e => {
            e.preventDefault();
            onRemove();
          }}>
            <FaTrash />
          </a>
        </OverlayTrigger>
      </td>
    </tr>
  );
};

export const NotesList = (props: NotesListProps) => {
  const { notes, onRemoveNote } = props;
  const [ removeNote, setRemoveNote ] = useState<Note>(null);
  const [ showModal, setShowModal ] = useState(false);

  return (
    <>
      <Table className={styles.table} striped bordered hover>
        <thead>
          <tr>
            <th>#</th>
            <th><FormattedMessage id="notes.list_table.columns.title" /></th>
            <th />
          </tr>
        </thead>
        <tbody>
          {notes.map((note, index) => (
            <NoteRow
              key={index}
              note={note}
              onRemove={() => {
                setShowModal(true);
                setRemoveNote(note);
              }}
            />
          ))}
        </tbody>
      </Table>
      <RemoveNoteModal
        show={showModal}
        note={removeNote}
        onClose={() => setShowModal(false)}
        onRemove={() => onRemoveNote(removeNote)}
      />
    </>
  );
};
