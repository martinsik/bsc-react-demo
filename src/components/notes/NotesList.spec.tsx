import React from 'react';
import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';
import { OverlayTrigger } from 'react-bootstrap';

import { NotesList } from './NotesList';
import { MockApp } from '../../spec/MockApp';

describe('NotesList', () => {
  let originalOverlayTriggerRender = OverlayTrigger.prototype.render;

  beforeAll(() => {
    originalOverlayTriggerRender = OverlayTrigger.prototype.render;
    OverlayTrigger.prototype.render = () => null;
  });

  afterAll(() => {
    OverlayTrigger.prototype.render = originalOverlayTriggerRender;
  });

  it('should renders without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render((
      <MockApp>
        <NotesList notes={[]} onRemoveNote={() => void 0} />
      </MockApp>
    ), div);

    ReactDOM.unmountComponentAtNode(div);
  });

  it('should render table line for every array item', () => {
    const mockData = [
      { id: 1, title: 'Title 1' },
      { id: 2, title: 'Title 2' },
      { id: 3, title: 'Title 3' },
    ];

    const component = TestRenderer.create((
      <MockApp>
        <NotesList notes={mockData} onRemoveNote={() => void 0} />
      </MockApp>
    ));
    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
