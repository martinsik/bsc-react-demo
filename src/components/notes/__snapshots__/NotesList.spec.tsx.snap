// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesList should render table line for every array item 1`] = `
<table
  className="table table table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th>
        <span>
          Title
        </span>
      </th>
      <th />
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
      </td>
      <td>
        <a
          href="/note/1"
          onClick={[Function]}
        >
          Title 1
        </a>
      </td>
      <td />
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        <a
          href="/note/2"
          onClick={[Function]}
        >
          Title 2
        </a>
      </td>
      <td />
    </tr>
    <tr>
      <td>
        3
      </td>
      <td>
        <a
          href="/note/3"
          onClick={[Function]}
        >
          Title 3
        </a>
      </td>
      <td />
    </tr>
  </tbody>
</table>
`;
